### PR TITLE
Intro `*Node` cache and recycling

### LINKED ISSUE #83: Add `idr.Node` pool and recycling to save repeated allocation
Some initial CSV reader prototype shows 10x reduction in memory alloc, and 2x reduction in latency (only 2x because recycling takes time to remove all nodes in the trees and putting them back to sync.Pool) -- still very much worth it!

### PR DESCRIPTION
**Pull request by jf-tech**
As a result, adding Release(*idr.Node) to FormatReader so that Ingester will free up *idr.Node allocations for recycling

Updated all benchmarks.

Not much degradation so far. The reason the json sample bench mark didn't improve much is because there isn't much nodes allocated per read. However typically in the CSV parsing/transform scenarios, files are usually very long and node allocation caching saving would start to show significantly.

resolves #83

### CODE REVIEW

**Comment by jf-tech** [comment]
all right i gave up. this PR it is. no checks that's fine. already ran tests many times locally.

File: customfuncs/javascript_test.go [hunk 1]
```diff
@@ -223,12 +223,12 @@ func TestJavascriptClearVarsAfterRunProgram(t *testing.T) {
 }
 
 // go test -bench=. -benchmem -benchtime=30s
-// BenchmarkIfElse-4                            	368017143	      98.1 ns/op	       8 B/op	       1 allocs/op
-// BenchmarkEval-4                              	26409430	      1386 ns/op	     418 B/op	       8 allocs/op
-// BenchmarkJavascriptWithNoCache-4             	  172803	    210958 ns/op	  136608 B/op	    1698 allocs/op
-// BenchmarkJavascriptWithCache-4               	23059004	      1572 ns/op	     129 B/op	       8 allocs/op
-// BenchmarkConcurrentJavascriptWithNoCache-4   	    1140	  32729941 ns/op	27328924 B/op	  339654 allocs/op
-// BenchmarkConcurrentJavascriptWithCache-4     	   70977	    504870 ns/op	   26568 B/op	    1745 allocs/op
+// BenchmarkIfElse-4                            	352522934	        97.9 ns/op	       8 B/op	       1 allocs/op
```
Comment (liangxibing) [inline customfuncs/javascript_test.go:226]: Looks not aligned with other lines

**Review by liangxibing — APPROVED** [review]
LGTM

File: customfuncs/javascript_test.go (hunk 1 above)
Comment (jf-tech) [reply customfuncs/javascript_test.go:226]: will fix later.

**Pull request merged**